Google.Ads.Gax/src/Util/ExceptionUtilities.cs

```diff
@@ -0,0 +1,197 @@
+            }
+            string messageToProcess = rpcException.Message.Substring(
+                prefixIndex + EXCEPTION_TYPE_PREFIX.Length);
+            string[] messageParts = messageToProcess.Split('\n');
+            if (messageParts.Length >= 0)
+            {
+                string lineToParse = messageParts[0];
+                string[] subParts = lineToParse.Split(',').Select(s => s.Trim()).ToArray();
+                // TokenErrorResponse is serialized as
+                // Error:\"{Error}\", Description:\"{ErrorDescription}\", Uri:\"{ErrorUri}\"
+                string error = "";
+                string description = "";
+                string uri = "";
+
+                foreach (string subPart in subParts)
+                {
+                    if (subPart.StartsWith(ERROR_PREFIX))
+                    {
+                        error = subPart.Substring(ERROR_PREFIX.Length).Trim('"');
+                    }
+                    else if (subPart.StartsWith(DESC_PREFIX))
+                    {
+                        description = subPart.Substring(DESC_PREFIX.Length).Trim('"');
+                    }
+                    else if (subPart.StartsWith(URI_PREFIX))
+                    {
+                        uri = subPart.Substring(URI_PREFIX.Length).Trim('"');
+                    }
+                }
+                return new TokenResponseException(new TokenErrorResponse()
+                {
+                    Error = error,
+                    ErrorDescription = description,
+                    ErrorUri = uri
+                });
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Parses the task exception.
+        /// </summary>
+        /// <param name="e">The <see cref="AggregateException"/> to parse.</param>
+        /// <returns>The parsed <see cref="AdsBaseException"/> if parsing is successful;
+        /// The underlying <see cref="RpcException" /> if the exception cannot be parsed as a
+        /// AdsBaseException.</returns>
+        /// <remarks><code>AggregateException</code> is very close to a catch-all exception for
+        /// Tasks. While all the situations that we know of involves this exception being thrown
+        /// by an underlying <code>RpcException</code>, theoretically this method may return a
+        /// null object, if the <code>AggregateException</code> is not caused by an
+        /// <code>RpcException</code>. That would typically indicate an underlying issue with
+        /// the code.</remarks>
+        internal static RpcException ParseTaskException<TResponse>(AggregateException e)
+        {
+            RpcException rpcException = ExtractRpcException(e);
+            AdsBaseException adsException = ParseRpcException<TResponse>(rpcException);
+            return (adsException == null) ? rpcException : adsException;
+        }
+
+        /// <summary>
+        /// Extracts the RPC exception from an <see cref="AggregateException"/>.
```

Comment:
Should mention that this returns the *first* RPC exception found, as an `AggregateException` could contain multiple `RpcException`s.

On that note, would it be better if this returned a collection of `RpcException` objects?

Reply:
I added clarifying comments.